### PR TITLE
Send email alerts on certain field changes.

### PR DESCRIPTION
**Pull request by awensaunders**
Sometimes there is a need to send emails when certain fields are changed and a record is assigned to multiple users (as in a referral). This PR is an implementation of that.

### CODE REVIEW

File: app/jobs/alert_notify_job.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+# Job for enqueuing notifications about changes to the User account
+class AlertNotifyJob < ApplicationJob
+  queue_as :mailer
+
+  def perform(alert_id)
+    alert = Alert.find(alert_id)
+    record = alert.record
+    users = record.associated_users
+    users.each do |user|
```
Comment (aespinoza-quoin) [inline app/jobs/alert_notify_job.rb:11]: If the job failed for one user, should continue with the next one? can you make sure that this is happening?

File: app/models/concerns/alertable.rb [hunk 2]
```diff
@@ -40,18 +49,36 @@ def remove_alert_on_save
...
+    alerts_on_change.each do |field_name, conf_record|
       next unless changed_field_names.include?(field_name)
 
-      add_alert(alert_for: FIELD_CHANGE, date: Date.today, type: form_name, form_sidebar_id: form_name)
+      add_field_alert(conf_record)
+    end
+  end
+
+  def add_field_alert(conf_record)
+    case conf_record.alert_strategy
```
Comment (aespinoza-quoin) [inline app/models/concerns/alertable.rb:71]: make sure you run rubocop for all the files added/modified
please leave a comment that this is partially implemented.

File: app/models/concerns/alertable.rb [hunk 3]
```diff
@@ -91,53 +119,77 @@ def add_approval_alert(approval_type, system_settings)
...
+  end
+
+  def open_enabled_records
+    joins(:alerts).where('data @> ?', { record_state: true, status: Record::STATUS_OPEN }.to_json)
+  end
+end
+
+# This class is used for the members of changes_field_to_form in system_settings.
+# It is used to store the form name, and the alert strategy (associated_users, owner, nobody)
+class AlertConfigEntry
```
Comment (aespinoza-quoin) [inline app/models/concerns/alertable.rb:182]: why not a new file?

File: app/services/alert_notification_service.rb [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  def record_type_translated
+    I18n.t("forms.record_types.#{record.class.parent_form}", locale:)
+  end
+
+  def form_section_name_translated
+    @form_section_name_translated ||= I18n.with_locale(locale) { form_section&.name }
+  end
+
+  def send_notification?
+    # We may want to add some checks here in the future.
```
Comment (aespinoza-quoin) [inline app/services/alert_notification_service.rb:53]: TODO?

File: app/mailers/record_action_mailer.rb [hunk 5]
```diff
@@ -42,6 +42,19 @@ def transfer_request(transfer_request_notification)
     mail(to: @transfer_request_notification&.transitioned_to&.email, subject: @transfer_request_notification.subject)
   end
 
+  def alert_notify(alert_notification)
+    @alert_notification = alert_notification
+    @user = @alert_notification.user
```
Comment (aespinoza-quoin) [inline app/mailers/record_action_mailer.rb:47]: @user, @locale_email and @record are not need it in the view you can use @alert_notification

File: app/views/record_action_mailer/alert_notify.html.erb [hunk 6]
```diff
@@ -0,0 +1,17 @@
+<!DOCTYPE html>
+<html lang="<%= @user.locale || I18n.locale %>">
```
Comment (aespinoza-quoin) [inline app/views/record_action_mailer/alert_notify.html.erb:2]: @alert_notification.locale

File: app/views/record_action_mailer/alert_notify.html.erb [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+  <head>
+    <meta content="text/html; charset=UTF-8" http-equiv="Content-Type" />
+  </head>
+  <body>
+    <p>
+      <%= t("email_notification.alert",
+          record_type: @alert_notification.record_type_translated,
+          id: link_to(@record.short_id, url_for_v2(@record)),
+          form_name: @alert_notification.form_section_name_translated,
+          locale: @locale_email
```
Comment (aespinoza-quoin) [inline app/views/record_action_mailer/alert_notify.html.erb:12]: @alert_notification.locale

File: app/views/record_action_mailer/alert_notify.text.erb [hunk 8]
```diff
@@ -0,0 +1,8 @@
+<%= t("email_notification.alert",
+    record_type: @alert_notification.record_type_translated,
+    id: @record.short_id,
+    form_name: @alert_notification.form_section_name_translated,
+    locale: @locale_email
```
Comment (aespinoza-quoin) [inline app/views/record_action_mailer/alert_notify.text.erb:5]: @alert_notification.locale

**Comment by aespinoza-quoin** [comment]
Please make sure to run rubocop.

**Comment by awensaunders** [comment]
> Please make sure to run rubocop.

Is there some particular way you’re invoking rubocop? I’m not seeing any rubocop warnings on these changed files.

**Comment by awensaunders** [comment]
I've added the ability to dismiss alerts by adding a new API route and some frontend stuff. I'm not super familiar with the frontend of primero, so I would appreciate any comments.

See video of what the frontend looks like below:
https://github.com/primeroIMS/primero/assets/3837160/31404b5e-2e25-45a8-9f18-1d3a52d3a778

File: app/javascript/components/records/action-creators.js [hunk 9]
```diff
@@ -168,6 +170,21 @@ export const fetchRecordsAlerts = (recordType, recordId, asCallback = false) =>
   }
 });
 
+export const deleteAlertFromRecord = (recordType, recordId, alertId) => ({
+  // TODO this should be a const interpolated here not a literal
```
Comment (jtoliver-quoin) [inline app/javascript/components/records/action-creators.js:174]: Remove todo, this should be fine as is

File: app/javascript/components/internal-alert/component.jsx [hunk 10]
```diff
@@ -30,13 +31,30 @@ const Component = ({ title, items, severity, customIcon }) => {
     [css.disableCollapse]: items?.size <= 1
   });
 
+  const renderDismissButton = handler => {
```
Comment (jtoliver-quoin) [inline app/javascript/components/internal-alert/component.jsx:34]: Make these renderItem and renderDismissButton components in another file under components/internal-alert/components

File: app/controllers/api/v2/alerts_controller.rb [hunk 11]
```diff
@@ -15,6 +15,18 @@ def index
     @alerts = @record.alerts
   end
 
+  def destroy
+    authorize! :update, @record
```
Comment (pnabutovsky) [inline app/controllers/api/v2/alerts_controller.rb:19]: Can you remove the :update permission check? The alert dismissal shows up on the show page, which means that someone without edit permissions but with remove alert permissions will be able to see the button. And this is a reasonable usecase to support: managers will likely want to dismiss alerts.

**Review by pnabutovsky — CHANGES_REQUESTED** [review]

**Pull request merged**